cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/writer/LineageWriterDatasetFramework.java

```diff
@@ -156,9 +156,26 @@ public void deleteAllInstances(Id.Namespace namespaceId) throws DatasetManagemen
     return dataset;
   }
 
+  @Nullable
+  @Override
+  public <T extends Dataset> T getDataset(Id.DatasetInstance datasetInstanceId, @Nullable Map<String, String> arguments,
+                                          @Nullable ClassLoader classLoader,
+                                          DatasetClassLoaderProvider classLoaderProvider,
+                                          @Nullable Iterable<? extends Id> owners, AccessType accessType)
+    throws DatasetManagementException, IOException {
+    T dataset = super.getDataset(datasetInstanceId, arguments, classLoader, classLoaderProvider, owners, accessType);
+    writeLineage(datasetInstanceId, dataset, accessType);
```

Comment:
I think the current implementation of `LineageWriterDatasetFramework` is not right for all `getDataset` calls. The contract of this method is that it returns `null` to indicate no such dataset instance. However, we never check if `T dataset` is null after calling super. This mean we are recording linage for non-existing dataset instance.

This also means we have no unit-test covering that.